### DISCUSSION TITLE
metadata from flac stream to HLS

### DISCUSSION THREAD

**Discussion by DennisHeerema** [savonet/liquidsoap#3282] [Q&A]
Hi There,

I use a flac icecast stream, catch that with liquidsoap and convert it to all my different lossy output formats.

With the conversion to HLS I loose my metadata.

This doesn't happen when converting to icecast streams (aac and mp3)
Also converting a mp3 stream to HLS works fine.
What am I doing wrong?

my current script:

```
log.file.path.set("C:/Program Files/liquidsoaphls/log/HLS.log")
log.stdout.set(true)
log.level := 4

url = "http://ip:80/studio"
input = mksafe(input.http(url))

s = input
	               
aac_lofi = %ffmpeg(format="adts", %audio(codec="aac", b="48k"))

aac_midfi =%ffmpeg(format="adts", %audio(codec="aac", b="64k"))

aac_hifi = %ffmpeg(format="adts", %audio(codec="aac"))

streams_info = [("aac_lofi",(48000,"mp4a.40.29","ts")), ("aac_midfi",(64000,"mp4a.40.2","ts")), ("aac_hifi",(220000,"mp4a.40.2","ts"))]

streams = [("aac_lofi",aac_lofi), ("aac_midfi", aac_midfi), ("aac_hifi", aac_hifi)]

def segment_name(~position,~extname,stream_name) =
  timestamp = int_of_float(time())
  duration = 10
  "#{stream_name}_#{duration}_#{timestamp}_#{position}.#{extname}"
end

output.file.hls(playlist="live.m3u8", segment_duration=10.0, segments=15, segments_overhead=10, segment_name=segment_name, "C:/nginx/html/hls", streams, input)

```

I also tryed converting to id3 format

```
aac_lofi = %ffmpeg(format="adts", %audio(codec="aac", b="48k")).{id3_version = 3}

aac_midfi =%ffmpeg(format="adts", %audio(codec="aac", b="64k")).{id3_version = 3}

aac_hifi = %ffmpeg(format="adts", %audio(codec="aac")).{id3_version = 3}
```

Kind regards,

Dennis

**Comment by toots** [comment]
Hi,

With `adts`, the metadata gets added as a initial id3 tag to each stream segment. Not all players will support it.

Things you can do:
* Try with `mpegts` instead
* Try different players:
  * VLC
  * hls.js demo page: https://hlsjs.video-dev.org/demo/

In particular, the HLS js demo will show you if metadata is being received, at least with mpegts.

My guess is that the metadata is here but not all clients support it.

**Comment by DennisHeerema** [comment]
Hi Thanks for the reply Toots,

Did some testing:
- When the source is a flac icecast stream 
both adts and mpegts do not play metadata

- When the source is a aac icecast stream
adts streams metadata
mpegst does not play metadata

Kind regards,

Dennis

**Reply by toots** [reply]
Ok. Where you able to test with vlc or hls.js demo page?

**Comment by DennisHeerema** [comment]
Tested with AIMP and VLC, both react equally.

Kind regards,

Dennis

**Reply by toots** [reply]
Just used this test script and was able to confirm metadata in vlc. You do need to have vlc read from http and not local files for it to work tho:

```liquidsoap
input = playlist("~/sources/test-station/playlists")

input = insert_metadata(input)

counter = ref(0)
def f() =
  counter := counter() + 1
  input.insert_metadata([("title", "test title #{counter()}"), ("album","foolol")])
end

thread.run(every=2., f)

aac_lofi = %ffmpeg(format="adts", %audio(codec="aac", b="48k"))

aac_midfi =%ffmpeg(format="adts", %audio(codec="aac", b="64k"))

aac_hifi = %ffmpeg(format="adts", %audio(codec="aac"))

streams_info = [("aac_lofi",(48000,"mp4a.40.29","ts")), ("aac_midfi",(64000,"mp4a.40.2","ts")), ("aac_hifi",(220000,"mp4a.40.2","ts"))]

streams = [("aac_lofi",aac_lofi), ("aac_midfi", aac_midfi), ("aac_hifi", aac_hifi)]

def segment_name(~position,~extname,stream_name) =
  timestamp = int_of_float(time())
  duration = 10
  "#{stream_name}_#{duration}_#{timestamp}_#{position}.#{extname}"
end

output.file.hls(fallible=true, playlist="live.m3u8", segment_duration=10.0, segments=15, segments_overhead=10, segment_name=segment_name, "/tmp/hls", streams, input)
```

**Reply by DennisHeerema** [reply]
yes i can confirm that VLC is doing meta data, that's not the point.
The metadata is missing when an flac icecast stream is used as source in liquidsoap to HLS.

When I use an icecast AAC stream as a source,  metadata is precent.

first screenshot VLC and script when flac stream is used

![afbeelding](https://github.com/savonet/liquidsoap/assets/3913639/29e30a05-4f97-43cd-9cc3-064bfad5bdcd)

second screenshot VLC and script when AAC is used

![afbeelding](https://github.com/savonet/liquidsoap/assets/3913639/bd99959c-4cfc-40ee-b2f4-ecdcc054d280)
kind regards,

Dennis

**Reply by toots** [reply]
Ha yes, there is a known limitation, which comes from the underlying ffmpeg engine/decoder. FFmpeg is seemingly unable to decode metadata in chained ogg/flac streams. We've got patches to fix the issue but they haven't made it upstream yet.

For lossless stream communication between liquidsoap instances I would recommend using another transport layer, for instance `output.srt`, and making sure that the ffmpeg decoder is either disabled or with a lower priority than the ogg decoder.

On windows, the SRT I/Os seem to be having some issues. In this case I would suggest reverting the calling setup and using `input.harbor` again, with the ffmpeg decoder either disabled or with a lower priority than the ogg decoder.

**Reply by DennisHeerema** [reply]
How do i lower or disable ffmpeg decoder?

found this: settings.decoder.priorities.ogg := 15
Do i put this on top of my script?

Kind regards,

Dennis

**Reply by toots** [reply]
Precisely, yes. And switch to `input.harbor` (which will receive the stream) instead of `input.http`.

**Reply by DennisHeerema** [reply]
Harber works fine!!
but the decodersetting gives an error:

```
C:\Program Files\liquidsoaphls>liquidsoap fzo.liq
At fzo.liq, line 3, char 0-27:
settings.decoder.priorities.ogg := 15

Error 5: this value has no method `ogg`
  Its type is
{
  aiff : (() -> int)
  .{comments : string, description : string, set : (int) -> unit
  },
  basic : (() -> int)
  .{comments : string, description : string, set : (int) -> unit
  },
  comments : string,
  description : string,
  ffmpeg : (() -> int)
  .{comments : string, description : string, set : (int) -> unit
  },
  srt : (() -> int)
  .{comments : string, description : string, set : (int) -> unit
  },
  wav : (() -> int)
  .{comments : string, description : string, set : (int) -> unit
  }
}.

C:\Program Files\liquidsoaphls>
```

**Reply by toots** [reply]
Yeah I forgot that you're using the windows build, which only has ffmpeg enabled.. In this case I wouldn't bother about the decoder and instead make sure to pass `send_icy_metadata=true` in your `output.icecast` sending to `input.harbor`.

**Reply by DennisHeerema** [reply]
the input harbor is a butt encoder, who creates the flac stream. So i do not think i can `put send_icy_metadata=true in your output.icecast sending to input.harbor.` anywhere...right?

So my config looks like this now, for those who are looking for examples:

```
log.file.path.set("C:/Program Files/liquidsoaphls/log/HLS.log")
log.stdout.set(true)
#settings.decoder.priorities.ogg := 15
#settings.decoder.priorities.flac := 1
# Input the stream,
# from an Icecast server or any other source


# url = "http://ip:80/studio"
# input = mksafe(input.http(url))


set("harbor.bind_addr","ip")
set("harbor.verbose",true)
input = input.harbor(
       "live",
       port=9000,
       password="Password"
)



aac_lofi = %ffmpeg(format="adts", %audio(codec="aac", b="48k"))

aac_midfi =%ffmpeg(format="adts", %audio(codec="aac", b="64k"))

aac_hifi = %ffmpeg(format="adts", %audio(codec="aac"))

streams_info = [("aac_lofi",(48000,"mp4a.40.29","ts")), ("aac_midfi",(64000,"mp4a.40.2","ts")), ("aac_hifi",(220000,"mp4a.40.2","ts"))]

streams = [("aac_lofi",aac_lofi), ("aac_midfi", aac_midfi), ("aac_hifi", aac_hifi)]

def segment_name(~position,~extname,stream_name) =
  timestamp = int_of_float(time())
  duration = 10
  "#{stream_name}_#{duration}_#{timestamp}_#{position}.#{extname}"
end

output.file.hls(fallible=true, playlist="live.m3u8", segment_duration=15.0, segments=15, segments_overhead=10, segment_name=segment_name, "C:/nginx/html/hls", streams, input)
```

When flac metadata to HLS is supported, i will combine this with my current script to this (to thinker during developement, i'll keep them seperated for now):
```

set("log.stdout",true)
set("log.file",false)
set("log.level",4)

# Input the stream,
# from an Icecast server flac stream
url = "http://ip:80/studio"
input = mksafe(input.http(url))

s = input
	                
# First transcoder: MP3 128 kbps
output.icecast(%mp3(stereo=true, stereo_mode="joint_stereo", bitrate=128, internal_quality=0), host = "ip", port = 80, password = "passwd", mount = "studio.mp3", genre="various pop hits oldies easylistening Nederlands",url="url",description="FlevoZiekenOmroep", id="FZO_MP3",name="FZO MP3 stream",icy_metadata="true",public=true, s)

# Second transcoder: AACPLUS 96 kbps 
output.icecast (%fdkaac(channels=2, samplerate=44100, bitrate=96, afterburner=false, aot="mpeg4_aac_lc", transmux="adts", sbr_mode=false), host = "ip", port = 80, password = "passwd", mount = "studio.aac", genre="various pop hits oldies easylistening Nederlands",url= "https://url",description= "FlevoZiekenOmroep", id="FZO_AACPLUS",name="FZO AACPLUS stream",icy_metadata="true",public=true, s)

aac_lofi = %ffmpeg(format="adts", %audio(codec="aac", b="48k"))

aac_midfi =%ffmpeg(format="adts", %audio(codec="aac", b="64k"))

aac_hifi = %ffmpeg(format="adts", %audio(codec="aac"))

streams_info = [("aac_lofi",(48000,"mp4a.40.29","ts")), ("aac_midfi",(64000,"mp4a.40.2","ts")), ("aac_hifi",(220000,"mp4a.40.2","ts"))]

streams = [("aac_lofi",aac_lofi), ("aac_midfi", aac_midfi), ("aac_hifi", aac_hifi)]

def segment_name(~position,~extname,stream_name) =
  timestamp = int_of_float(time())
  duration = 10
  "#{stream_name}_#{duration}_#{timestamp}_#{position}.#{extname}"
end

output.file.hls(fallible=true, playlist="live.m3u8", segment_duration=15.0, segments=15, segments_overhead=10, segment_name=segment_name, "C:/nginx/html/hls", streams, input)
```

So everything in one script. is my goal.

Kind regards
